### PR TITLE
[OSPP] Morphological Transformations

### PR DESCRIPTION
**Pull request by Ris-Bali**
The PR aims to implement morphological transformations. A PR for #70.

### CODE REVIEW

File: include/Utils/DIPUtils.h [hunk 1]
```diff
@@ -570,6 +675,122 @@ Value zeroCond(OpBuilder &builder, Location loc, Type elemType,
...
+// Utility function for morphological operations ; cannot handle tail processing
+void calcAndStorewoTailProcessingMorph(OpBuilder &builder, Location loc,
+                                       VectorType vecType, Value inputVec,
+                                       Value kernelVec, Value output,
+                                       Value beginIdx, Value endIdx,
+                                       Value zeroPadding,
+                                       Value inputCol, VectorType vectorMaskTy,
+                                       Type elemTy, Value kernelValue,
+                                       Value zeroPaddingElem, DIP_OP op) {
+
```
Comment (meshtag) [inline include/Utils/DIPUtils.h:714]: Remove this space.

**Review by meshtag — CHANGES_REQUESTED** [review]
Next review iteration.

File: include/Utils/DIPUtils.h [hunk 2]
```diff
@@ -645,10 +866,27 @@ void traverseImagewBoundaryExtrapolation(
                 Value inputVec =
                     builder.create<BroadcastOp>(loc, vectorTy32, constantValue);
 
+                Value tailCond =
```
[the remark this answers is not in the document]
Comment (Ris-Bali) [reply include/Utils/DIPUtils.h:869]: Yes there is no need for this condition.

**Comment by meshtag** [comment]
I do not think we require tests with multiple type inputs/outputs for these operations, the type handling infrastructure is same across all operations in the dialect and hence can be verified with type specific tests only for `correlation_2d` op. Having them for all other operations as well feels like a lot of avoidable code duplication to me.

File: docs/dip-opt.md [hunk 3]
```diff
@@ -109,3 +109,236 @@ Following schematic depicts the process(with boundary extrapolation and variable
...
 and a 3x3 kernel
 ![](./Images/AnchorPointAndBoundaryExtrapolation.png)
+
+### 2. 2D Erosion(erosion_2d)
+
+dip.erosion_2d performs boundary extrapolation for making the size of output image
+equal to the size of input image and then uses coefficient broadcasting and
+strip mining(CBSM) approach for performing erosion.
+
+Boundary extrapolation can be done using
```
Comment (meshtag) [inline docs/dip-opt.md:119]: No need to copy paste this for each and every op in morphological operations category. 

Create a separate section for morphological operations and then explain each operation in respective subsections.

**Comment by meshtag** [comment]
Ping me once you fix the issue with examples

**Review by meshtag — APPROVED** [review]

**Pull request merged**